### ISSUE TITLE
Add location to rack clone fields

### ISSUE THREAD

**Issue by alehaa** [netbox-community/netbox#16786]
### NetBox version

v4.0.6

### Feature type

Change to existing functionality

### Proposed functionality

When creating or editing a device, all racks in a location are selectable, including racks in child locations. If you select a rack in a child location, saving the device is rejected with this error message:

> Rack RACK01 does not belong to location PARENT.

Therefore, I suggest that all racks of a location (including child locations) can be selected in the form. When saving, the location of the device will be replaced by the location of the rack, thus automatically correcting it if needed.

**Steps to reproduce:**

* Create location PARENT
* Create child location CHILD
* Create rack RACK01 in CHILD
* Create new device. Select location PARENT and RACK01 as rack.
* Save

### Use case

The structure of the locations is entirely up to the NetBox user. A typical scenario could be to have parent locations for the room and child locations for the rack rows named "R1", "R2", ... in each room. Searching for the room is quite easy because it is unique. Finding the right rack row is not as easy, as there are several with the same name.

Also, it seems a bit odd that the racks are selectable in the UI and the backend does not accept the user's choice.

### Database changes

None

### External dependencies

None

**Comment by alehaa** [comment]
I volunteer to work on a PR for this feature.

**Comment by jeremystretch** [comment]
> Searching for the room is quite easy because it is unique. Finding the right rack row is not as easy, as there are several with the same name.

**Comment by alehaa** [comment]
No, the (device) objects are pretty easy for us to find. However, creating / editing them sometimes has pitfalls.

Real life problem on our site:
1. Employee creates new device.
2. Employee assigns device to location `Data Center Room 01`, rack `Rack 17`.
4. Operator saves the device.
5. NetBox displays error because `Rack 17` is not in `Datacenter Room 01`, but in `Datacenter Room 01 > Row A`.

This problem occurs because the Device edit form displays all racks in a location, including child locations. There are (at least) two possible solutions:
1. Limit the racks in the device edit form to those of the location, NOT including child locations. This would reflect the implementation of `dcim.models.Device.check()`.
2. Change the implementation of `Device.check()` / `Device.save()` so that the location (and perhaps the site) is copied from the rack. Since the rack has a defined location / site, its clear where the device should be located.

However, my inquiry was that solution 1 could make the workflow a bit difficult for employees: While it is clear to search for the location `Data Center Room 01`, searching for `Row A` might result in multiple items from different locations named the same. The device edit form doesn't display additional information about a location's parent locations. Also, a technician might not even know what row a rack is in, only the room (parent location). Typing the room into the location field of the device edit form wouldn't show the child locations, so you might be able to select them if you only know the parent location.
